File: solutions/clojure/bob/1/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn is-question? [s]
  (-> s str/trim (str/ends-with? "?")))

(defn is-yell? [s]
  (and (= s (str/upper-case s))
       (boolean (re-find #"\p{L}" s))))

(defn is-yell-and-question? [s]
  (and (is-question? s) (is-yell? s)))

(defn response-for [s]
  (cond
    (str/blank? s) "Fine. Be that way!"
    (is-yell-and-question? s) "Calm down, I know what I'm doing!"
    (is-question? s) "Sure."
    (is-yell? s) "Whoa, chill out!"
    :else "Whatever."))

File: solutions/java/bob/1/src/main/java/Bob.java
class Bob {
    String hey(String param) {
        param = param.trim();

        String result = "Whatever."; // presume anything else

        boolean isQuestion = param.endsWith("?");
        boolean isYell = param.toUpperCase().equals(param) && hasAlpha(param);
        boolean isEmpty = param.isEmpty();

        if(isEmpty){
            result = "Fine. Be that way!";
        }else if (isQuestion && isYell) {
            result = "Calm down, I know what I'm doing!";
        } else if (isQuestion) {
            result = "Sure.";
        } else if (isYell) {
            result = "Whoa, chill out!";
        }

        return result;
    }

    boolean hasAlpha(String s){
        if (s == null) {
            return false;
        }

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z')) {
                return true;
            }
        }
        return false;
    }
}